File: LeetCode - Algorithm/Hash/128.py
# 128. Longest Consecutive Sequence
# https://leetcode.com/problems/longest-consecutive-sequence/

from typing import List


class Solution:
    # This approach obviously has a time complexity of O(n^2)
    def longestConsecutive1(self, nums: List[int]) -> int:
        numsSet = set(nums)

        result = 0

        for num in numsSet:
            target = num
            currentResult = 1

            while target + 1 in numsSet:
                currentResult += 1
                target += 1

            result = max(result, currentResult)

        return result

    # To improve it, I brought a new approach that uses a hash map to reduce the number
    # of recursive calls, but it still results in O(n^2) time complexity
    def longestConsecutive2(self, nums: List[int]) -> int:
        numsSet = set(nums)
        hash = {}

        result = 0

        for num in numsSet:
            target = num
            currentResult = 1

            while target + 1 in numsSet:
                if target + 1 in hash:
                    currentResult += hash[target + 1]
                    break

                currentResult += 1
                target += 1

            hash[num] = currentResult

            result = max(result, currentResult)

        return result

    # By excluding elements that are not the start of a consecutive sequence, this final
    # approach achieves a time complexity of O(n).
    def longestConsecutive3(self, nums: List[int]) -> int:
        numsSet = set(nums)

        result = 0

        for num in numsSet:
            if num - 1 in numsSet:
                continue

            target = num
            currentResult = 1

            while target + 1 in numsSet:
                currentResult += 1
                target += 1

            result = max(result, currentResult)

        return result

File: LeetCode - Algorithm/Matrix/128.js
// 128. Longest Consecutive Sequence
// https://leetcode.com/problems/longest-consecutive-sequence/

/**
 * @param {number[]} nums
 * @return {number}
 */
var longestConsecutive = function (nums) {
  const numsSet = new Set(nums);

  const targets = [];

  for (let i = 0; i < nums.length; i++) {
    if (numsSet.has(nums[i] + 1)) continue;

    targets.push(nums[i]);
  }

  let result = 0;
  for (let i = 0; i < targets.length; i++) {
    let currentResult = 0;
    let numToCheck = targets[i];
    while (numsSet.has(numToCheck)) {
      currentResult += 1;
      numToCheck -= 1;
    }

    result = Math.max(result, currentResult);
  }

  return result;
};

// I should remember that I can't select a number from a set using indexing.
// And, it's annoying that same logic passed in Python but failed in JS. :(

/**
 * @param {number[]} nums
 * @return {number}
 */
var longestConsecutive = function (nums) {
  const numsSet = new Set(nums);

  let result = 0;
  for (let i = 0; i < nums.length; i++) {
    if (numsSet.has(nums[i] + 1)) continue;

    let currentResult = 0;
    let currentTarget = nums[i];
    while (numsSet.has(currentTarget)) {
      currentResult += 1;
      currentTarget -= 1;
    }

    result = Math.max(result, currentResult);
  }

  return result;
};

/**
 * @param {number[]} nums
 * @return {number}
 */
var longestConsecutive = function (nums) {
  const numsSet = new Set(nums);

  let result = 0;
  for (let i = 0; i < nums.length; i++) {
    if (numsSet.has(nums[i] + 1)) continue;

    let length = 1;
    while (numsSet.has(nums[i] - length)) {
      length += 1;
    }

    result = Math.max(result, length);
  }

  return result;
};

/**
 * @param {number[]} nums
 * @return {number}
 */
var longestConsecutive = function (nums) {
  const numsSet = new Set(nums);

  let result = 0;
  for (let i = 0; i < nums.length; i++) {
    if (numsSet.has(nums[i] - 1)) continue;

    let length = 1;
    while (numsSet.has(nums[i] + length)) {
      length += 1;
    }

    result = Math.max(result, length);
  }

  return result;
};

/**
 * @param {number[]} nums
 * @return {number}
 */
var longestConsecutive = function (nums) {
  const numsSet = new Set(nums);

  let result = 0;
  for (let num of nums) {
    if (numsSet.has(num - 1)) continue;

    let length = 1;
    while (numsSet.has(num + length)) {
      length += 1;
    }

    result = Math.max(result, length);
  }

  return result;
};
